### PR TITLE
Reappeal ess-r-source-mode to make xref useful (was #475 xref backend for R)

### PR DESCRIPTION
**Pull request by dickmao**
From #475 

> > If you just remove the ess-r-source-mode I think this is good to merge!
> 
> I have removed this code, as you both requested. However, I'll make one final plea for its inclusion:

Not only do I see a real difference between `ess-r-source-mode` and `ess-dump-object-info-edit-buffer` in both function and intent, I also don't see how they could share code.  My experience has been that @atheriel's original PR works exactly how I'd want it -- the source viewing buffers are read-only and buried via `q`.  I can `M-.` and `M-,` to my heart's content.  

Other changes include a preliminary `test_that` for ESSR and setting `ess-r-package-library-paths` to a `libPaths()` call.

### CODE REVIEW

File: lisp/ess-r-flymake.el [hunk 1]
```diff
@@ -163,9 +163,13 @@ into a list and call REPORT-FN on it."
        for msg = (match-string 4)
        for (beg . end) = (let ((line (string-to-number (match-string 1)))
                                (col (string-to-number (match-string 2))))
-                           (flymake-diag-region src-buffer line col))
+                           (if (fboundp 'flymake-diag-region)
```
Comment (lionel-) [inline lisp/ess-r-flymake.el]: Could you make unrelated changes in other PRs please?

File: lisp/ess-r-mode.el [hunk 2]
```diff
@@ -2312,6 +2311,44 @@ state.")
...
+  (let ((inhibit-read-only t))
+    (setq header-line-format
+          (concat "Environment: " 
+                  (ess-r-source--parse-environment (current-buffer))
+                  (and (ess-r-source--parse-byte-compiled-p (current-buffer))
+                       " (byte-compiled)")))))
+
+(defun ess-r-source--parse-environment (buff)
+  "Determine the environment of the R function printed in BUFF.
+
```
Comment (lionel-) [inline lisp/ess-r-mode.el]: We don't use newlines after the first line. Can you please remove all of them?

File: lisp/ess-custom.el [hunk 3]
```diff
@@ -1501,6 +1501,9 @@ line starts with the roxy prefix.")
 
 ;; SJE -- this should not be defcustom - user does not set it.
 (defvaralias 'ess-current-process-name 'ess-local-process-name)
+
+;; should this have been make-buffer-local-variable?
+;; we always setq-local `ess-local-process-name' per buffer
```
Comment (lionel-) [inline lisp/ess-custom.el]: Since you are not a core developer, can you ask questions in an issue or github comment rather than in the source code?

**Comment by dickmao** [comment]
> Thanks. This is well written code...

Much of it is @atheriel 

> With this change, xref basically becomes useless for me.

I don't doubt your claim, but I don't know why.  `master` is only practical for your workflow of xref'ing your own package (it is not practical for the general case of inspecting outside packages).  I tried to preserve that.

> R is likely going to enable srcrefs by default and the long term plan is to reconstruct package directories from those, when local sources are available.

I don't doubt that claim, but I'd rather not wait for it.  Before I got into/created the packaging mess, I spent a couple days poring over the (very verbose)  discussion in #475 and #553 .  I tried to make the srcref chicanery work (and yes, chicanery is the right word as you know from #553).  It'd be hard work to tastefully get the user to configure all the right variables to download the srcrefs for *all* the tokens he might `M-.`  It would result in duplicated sources (one for the original libraries in /usr and one for the downloaded keep-source libraries in his home directory).  But worst of all, it results in the same outcome as dumping the function definitions.  In other languages, you see the context of the meta-dotted function.  In R, you just see the function (*with comments*) because R convention is to write one function per file (like matlab).

> You are introducing a new major mode instead of using the dump-object mechanism. As we have discussed it last time, we think the existing mechanism should be used, and perhaps improved.

Yes, I went over this argument with @atheriel in #475.  I can only regurgitate what he said -- that to a pair of fresh eyes the "edit function" and "read function" are sufficiently different in function and intent, and that the `ess-r-source-mode` which he derived from the emacs `special-mode` -- a mode specifically designed for quick one-off reads -- is highly apt.

Now I am guilty of the same verbosity I dreaded in #475.  The xref functionality is much desired by any ESS user who knows about `M-.`  This is probably a low percentage but still numbers in the hundreds.  As it stands, the configuration demands on the user are way too high for any kind of adoption.

**Comment by lionel-** [comment]
> is only practical for your workflow of xref'ing your own package 

It is practical for any package you have installed locally. Long term plan is to make it practical for all packages.

>  get the user to configure all the right variables to download the srcrefs for all the tokens

I don't understand what you mean.

> R convention is to write one function per file

I am not aware of such a convention.

> the "edit function" and "read function" are sufficiently different in function and intent

I disagree. There's no advantage in making the dump buffer artificially read-only. Besides what's the advantage over just printing the function in the console? More importantly, xref is for jumping to source location, not for dumping into a standalone buffer.

**Comment by dickmao** [comment]
Okay, I see your position hasn't change since May.  I am closing this PR.

**Comment by mmaechler** [comment]
> > R convention is to write one function per file
> 
> I am not aware of such a convention.

It is 100%  non-sense (put *politely* !) !     ... and don't go for "the average package" -- rather e.g. base R, recommended packages, other pieces which have stood the time  --  but even with the average package I guess,  your claim would be  wrong by orders.

@dickmao : We can't even know who you are (as you've decided to remain anonymous on GH).
What are your credentials  in Statistics and in R ?
Reading  statements  as the one above (plus an attitude of being radical and pretty carefree instead of considerate and  cautious) make me wonder...

I'm sorry I've felt to be blunt... but I'd rather enjoy my work, including on ESS., so I do hope you can adapt to how we collaborate here.

**Comment by atheriel** [comment]
@dickmao FWIW, although I'm flattered that you agree with me, ultimately I think that @lionel- and the other maintainers deserve our deference in what is or is not going to become part of ESS. Making decisions is a key part of their role, after all.

A more constructive route here would have been to open an issue asking if the views expressed in earlier discussion had changed before embarking on a PR. In this case, maybe this feature could be released as a separate package first to gauge interest.

> if you hypothetically suspended your biases

Some unsolicited advice: I have **never** seen anything good come of attacking the character of other developers. It's hard not to feel hurt when your contributions (or arguments) are rejected, but that's not a good reason to lash out.

**Comment by dickmao** [comment]
@atheriel Thank you for trying to make `xref` happen.

In the week that I've been involved in the project I don't think I've said anything unprofessional, certainly not commensurate to the comments levelled against me.  The line regarding biases I concede is borderline.

**Comment by vspinu** [comment]
As an observer from the inner side, I back it up, you haven't said anything unprofessional or attacked anyone personally. You are just not a good salesman. You want too many things and all at once. :wink:. Anyways, back to the PR ...

As always there are many unrelated to the main issue fixes which are good. I would like to cherry pick and pull those like last time. Would you mind? 

Regarding the core issue @lionel- pointed pretty much everything. I would just add here that that hole was left intentionally as a commitment device to all of us. First, by not having 100% functional xref users are forced to discover `ess-r-package-library-paths` which otherwise might be simply unnoticed. Second, frustrated devs like you (and me and @lionel- and ...) will bring PRs which will eventually land in master and do something much more useful than a simple `dump`. 

Till @lionel-  brings along his reconstruction idea to life (I still think it's pretty crazy, but well, it's his job to prove me wrong :stuck_out_tongue_winking_eye: ) we can do fetching CRAN and git archives into a dedicated directory and unpack those there (also @lionel-'s original idea if I am not mistaken). If you want to help with xref this is where you can really help.

Even more, I have recently pushed through AVFS support to CIDER (see end tip [here](https://cider.readthedocs.io/en/latest/interactive_programming/)) and I couldn't be happier with it. In a nutshell, on a xref you land inside a jar archive and all emacs works there as in ordinary file system (including non-trivial stuff like projectile, grep etc). Something along those lines would be nice here as well, but not a big deal though.

**Comment by dickmao** [comment]
> I would like to cherry pick and pull those like last time. Would you mind?

That'd be swell, and appreciate your vouching for my online behavior.
